packages/@ember/application/instance.d.ts

```diff
@@ -0,0 +1,37 @@
+import { EngineInstanceOptions, Factory, LookupOptions } from '@ember/-internals/owner';
+import { EngineInstance } from '@ember/engine';
+
+export class BootOptions {
+  isBrowser: boolean;
+  shouldRender: boolean;
+  document: Document | null;
+  rootElement: string | Element | null;
+  location: string | null;
+
+  constructor(options?: {
+    isBrowser?: boolean;
+    shouldRender?: boolean;
+    document?: Document;
+    rootElement?: string | Element;
+    location?: string;
+    isInteractive?: boolean;
+  });
+
+  toEnvironment(): Record<string, unknown>;
+}
+
+export default class ApplicationInstance implements EngineInstance {
+  lookup<T>(fullName: string, options?: LookupOptions): T | undefined;
+  factoryFor<T, C>(fullName: string, options?: LookupOptions): Factory<T, C> | undefined;
```

Comment:
This has the same issue. Basically, the rule of thumb is: if the generic is only used once, it's wrong in some way—because it's not actually constraining anything. It is the type system equivalent of an unused parameter to a function.

```suggestion
  factoryFor(fullName: string, options?: LookupOptions): Factory<unknown, unknown> | undefined;
```

Reply:
There's a constraint in `Factory`.

Reply:
Because it is in the return position, it will not be inferred (it cannot be!) and therefore it acts unconstrained.

Reply:
We’ll I removed it anyway.